torch/distributed/_shard/checkpoint/metadata.py

```diff
@@ -77,3 +78,34 @@ class TensorReadRequest:
     # offset and length w.r.t. to the storage identified by ``storage_key``
     offsets: Tuple[int, ...]
     lengths: Tuple[int, ...]
+
+
+@dataclass(frozen=True)
+class MetadataIndex:
+    """
+    This class represents a lookup key for items in a state dict or Metadata.
+    """
+    fqn: str
+    """Fully Qualified Name of the object"""
+
+    offset: Optional[torch.Size] = None
+    """If the object is a tensor, offset into the tensor we're looking for"""
+
+    index: Optional[int] = field(hash=False, compare=False, default=None)
+    """
+    Index hint when searching for tensor chunk to speedup lookups (optional)
+
+    A common representation of a sharded tensor is as a list of chunks so to
+    find the index in such a list you need to linear search it.
+
+    When constructing an instance of MetadataIndex that points to that list,
+    one can provide the index as a hint and it will be probed first before
+    the linear search and thus making it significantly faster.
+    """
+
+    def __init__(self, fqn: str, offset: Optional[Sequence[int]] = None, index: Optional[int] = None):
```

Comment:
why we need this init function given that it's a frozen dataclass? Maybe we can get rid of it?

Reply:
We need this because we need to ensure `offset` is a `torch.Size` consistently otherwise we need to keep constructing temporary instances all the time when doing comparisons as `torch.Size([0]) != [0]`.

This `__init__` makes it possible to pass a list or tuple so this works: `MetadataIndex("foo", [1]) == MetadataIndex("foo", torch.Size([1]))`.

It's big quality of life help that cleaned up a lot of gulk conversions in my prototype.